torch/_decomp/decompositions.py

```diff
@@ -848,62 +848,61 @@ def addmm(self: Tensor, mat1: Tensor, mat2: Tensor, beta: int = 1, alpha: int =
         return out
     return beta * self + out
 
+# This computes the mean and variance along the specifized normalization dims,
+# then normalizes along those dims. Finally, it returns the mean and variance of
+# the normalized dims. Note that it intentionally leaves outputs upcasted.
+# Example:
+# input: [2, 3, 4, 5], norm_dims: [1, 3]
+# mean: [2, 1, 4, 1]
+def normalize(input, norm_dims, eps):
+    computation_dtype = utils.get_computation_dtype(input.dtype)
+    input_acc = input.to(dtype=computation_dtype)
+    biased_var = torch.var(input_acc, dim=norm_dims, unbiased=False, keepdim=True)
```

Comment:
note that current nvfuser integration doesn't understand var, so if you want decomposition that can be fused today, it has to be sad `mean = ...; var = (x-mean)*(x-mean)`. Incidentally, that's also what Triton does in their manually written layer_norm.

Reply:
Can’t we just decompose var separately?

I actually originally wrote this with torch.var_mean, but it seemed like there were some minor numerical differences compared to calling var and mean separately.

Reply:
We could decompose `var` separately but it would be inefficient as `var` decomposition can be recomposed by backends as 2-pass algorithm at best, whereas computing `var` directly is one pass (that triton can't do, lol, because Welford algorithm is horrible). It doesn't matter too much for layer norm, because layer norm typically fits in L1 (Triton's new implementation) or registers (Triton's old implementation), so 1- or 2-pass doesn't matter. 
It's weird that var_mean and `var` and `mean` separate calls produce different values, they literally call the same kernel - do you have an example?

Reply:
BTW, I'd quite like it for meta function coverage reasons if we had a var_mean decomp :)

Reply:
@Chillee did you forget to account for Bessel correction?

Reply:
Normalization happens without Bessel correction (hence "biased_var")

Reply:
@ngimel it's actually a difference in the `mean` part of the computation, and only on cuda.

```
import torch

x = torch.randn(5000, 5000, device='cuda')
orig = torch.var_mean(x, dim=0)
decomp = (torch.var(x, dim=0), torch.mean(x, dim=0))

print((orig[1] - decomp[1]).abs().max())
```

It's possible there's no meaningful difference, and I should just relax the tolerances in the testing.

Reply:
Ah ok, `mean` being different I can believe (it uses different vectorization/tiling, because just mean is not as register-hungry as mean_var), won't fix, relax the tolerance.